feat: avoid storing empty/no-change caches + add toggle to skip cache store

johanblumenberg opened the pull request:
## Description

### Fix existing tests

The `run()` function in `src/restore.ts` was run twice when a test is run. Once because `lib/restore` is imported into the test so it runs automatically, and once because the restore file is executed from the test.

Removed the wait commands. I think those were there to wait for the `run()` function which is imported through `lib/restore` to complete. Removing the import removes the need to wait.

Merged the two tests into one test. The save action requires that the restore action was already run once, othewise it doesn't find the buildcache binary. I think it is better to execute the restore code in the test, and avoid importing `lib/restore`. This is also doing the same thing as a real build would do, instead of running the save code in a "fake" restored state in the tests.

### Avoid storing the cache in some situations

Avoid storing the cache if it is empty. This might be because the job failed to download the cache, or because the job was cancelled before downloading the cache. In this case, creating a new empty cache will ruin it for the next job.

Avoid storing the cache if there was no change. This improves the run time of the job and it also avoids storing multiple identical caches.

Add option to skip storing of caches. This can be used in jobs that are only run on PR branches, where the cache will never be used by anyone.

mikehardy reviewed (changes requested):
I love this idea, sincerely, but I don't like that it adds a new feature, a fair bit of output parsing etc, and instead of increasing testing it decreases the tests available
I use things like buildcache because I have negative amounts of time available for anything, and looking to the future I rely incredibly heavily on tests passing to indicate I can ingest a change correctly and release without having some massive time-suck of a regression issue

I really need those tests to not be disabled and for the new functionality to have some coverage in testing

specifically -
- it appears to parse the output format of buildcache with a regex? Classic future bug at some point where we've forgotten this feature is there, and it will just fail silently unless there's a test case where we send *actual* output of a real buildcache run through it to make sure it's working by verifying actual v expected
- does the cache store every time when skip is unspecified and when it is specified but false?
- does the cache skip no matter what if skip is specified and false?

I'd love to merge this but I need to know that skip is actually working and I need to know that a year from now when buildcache changes output format that we will at least be aware we have an error now. I hope you understand :pray:  - add some test cover and we're good to go

johanblumenberg commented:
Hi @mikehardy,

I fully agree with you regarding test coverage. If you think it is a good idea to add these features, I will add tests. I commented out the tests because they were failing, and I didn't want to spend time on that if the PR was not going to be accepted anyway.

I don't have any experience of writing tests for github actions, so this is a new area for me. A few pointers would be appreciated :)
- Would it be OK to mock `@actions/cache`? Then it would be easy to test the `save()` function in isolation in a unit test, and verify if the `cache.saveCache()` method is invoked or not.
- It should be quite easy to test the `printStats()` function in isolation, to make sure that the parsing works. Any idea how I can make `buildcache` return different stats, not to only test with zero stats?

mikehardy commented:
I just did some build performance / cache work recently so I'm recently sensitized to useless cache stores myself :smile: 

- https://github.com/ankidroid/Anki-Android/pull/11032
- https://gist.github.com/mikehardy/f076172f3b28826898c55dc03fe202de

So I think the feature is worthwhile, yes

- I think we may assume that actions/cache is working, so mocking sounds right, then you don't have to worry about external stuff, except the buildcache binary itself
- unsure how to make buildcache return different stuff, I suppose you'd have to set up a specific cache directory for test that was clean/empty (and maybe run the clear + zero stats commands to verify) then do some trivial build to make sure first run was all miss, then run same trivial build to make sure it was all hit, and both were parsed?

Brainstorming warning, all ideas highly suspect. Your initial code was great though except my plea for testing ;-), so I'm guessing whatever you come up with will make sense...

Commit:
Fix existing tests

The run() function in `src/restore.ts` was run twice when a test
is run. Once because `lib/restore` is imported into the test so
it runs automatically, and once because the restore file is
executed from the test.

Removed the wait commands. I think those were there to wait for
the run() function which is imported through `lib/restore` to
complete. Removing the import removes the need to wait.

Merged the two tests into one test. The save action requires
that the restore action was already run once, othewise it
doesn't find the buildcache binary. I think it is better to
execute the restore code in the test, and avoid importing
`lib/restore`. This is also doing the same thing as a real
build would do, instead of running the save code in a "fake"
restored state in the tests.

Commit:
Avoid storing the cache in some situations

Avoid storing the cache if it is empty. This might be because the
job failed to download the cache, or because the job was cancelled
before downloading the cache. In this case, creating a new empty
cache will ruin it for the next job.

Avoid storing the cache if there was no change. This improves the
run time of the job and it also avoids storing multiple identical
caches.

Add option to skip storing of caches. This can be used in jobs that
are only run on PR branches, where the cache will never be used by
anyone.

johanblumenberg commented on __tests__/main.test.ts:
```diff
@@ -3,76 +3,172 @@ import * as process from 'process'
 process.env['RUNNER_TEMP'] = tempPath
 process.env['RUNNER_TOOL_CACHE'] = cachePath
 process.env['GITHUB_WORKSPACE'] = ghWorkspace
+process.env['BUILDCACHE_DIR'] = buildcacheDir
```
Set `BUILDCACHE_DIR` to a folder inside the workspace, to automatially wipe the cache between tests

johanblumenberg commented on __tests__/main.test.ts:
```diff
@@ -3,76 +3,172 @@ import * as process from 'process'
 import * as cp from 'child_process'
 import * as path from 'path'
 import * as rimraf from 'rimraf'
-import { downloadLatest, install } from '../src/restore'
```
Remove the import of `restore`, to avoid that it automatically executes the `run()` function, which the test has no control over. It is also conflicting with later invocations of the restore action, because those will find that the symlinks are already set up and fail.

mikehardy reviewed (approved):
wow, now *that* is pretty thorough

mikehardy commented:
prettier check tripped up on something :thinking:

johanblumenberg commented:
> prettier check tripped up on something 🤔

Fixed it

mikehardy commented:
@johanblumenberg I'm really sorry this is such a slow process, I've been busy, and I really appreciate your patience. Looks like the build-test / build action is failing now :thinking: https://github.com/mikehardy/buildcache-action/runs/7046648693?check_suite_focus=true

johanblumenberg commented:
> @johanblumenberg I'm really sorry this is such a slow process, I've been busy, and I really appreciate your patience. Looks like the build-test / build action is failing now 🤔 https://github.com/mikehardy/buildcache-action/runs/7046648693?check_suite_focus=true

It runs OK when I run locally. I suspect that the problem might be that when running on github it will download the cache from the previous run, so it's not always running on a clean cache. But I haven't had time to look into it.

johanblumenberg commented:
@mikehardy It looks like this is the failue:

```
    ENOENT: no such file or directory, mkdir '/home/runner/work/buildcache-action/buildcache-action/__tests__/runner/WORKSPACE'
      25 |   rimraf.sync(ghWorkspace)
      26 |
    > 27 |   fs.mkdirSync(ghWorkspace)
         |      ^
      28 | })
```

The `__tests__/runner` folder doesn't exist in a clean clone. I changed it to recursively create the folder.

mikehardy commented:
Fantastic! developer to developer I feel like this was more arduous than it needed to be, with the testing issues prior to your work and then the slow responses - apologies for that but thanks for sticking with it. Packing this up for a release now!
